### PR TITLE
A3-1-5: Remove extraneous cases and handle template instantiations

### PR DESCRIPTION
**Pull request by lcartey**
## Description

Fixes https://github.com/github/codeql-coding-standards/issues/382 and https://github.com/github/codeql-coding-standards/issues/481.

## Change request type

 - [ ] Release or process automation (GitHub workflows, internal scripts)
 - [ ] Internal documentation
 - [ ] External documentation
 - [x] Query files (`.ql`, `.qll`, `.qls` or unit tests)
 - [ ] External scripts (analysis report or other code shipped as part of a release)

## Rules with added or modified queries

 - [ ] No rules added
 - [ ] Queries have been added for the following rules:
     - _rule number here_
 - [x] Queries have been modified for the following rules:
     - `A3-1-5`

## Release change checklist

A change note ([development_handbook.md#change-notes](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#change-notes)) is required for any pull request which modifies:

 - The structure or layout of the release artifacts.
 - The evaluation performance (memory, execution time) of an existing query.
 - The results of an existing query in any circumstance.

If you are only adding new rule queries, a change note is not required.

_**Author:**_ Is a change note required? 
  - [x] Yes
  - [ ] No

🚨🚨🚨
_**Reviewer:**_ Confirm that format of *shared* queries (not the .qll file, the
.ql file that imports it) is valid by running them within VS Code. 
  - [x] Confirmed

_**Reviewer:**_ Confirm that either a change note is not required or the change note is required and has been added.
  - [ ] Confirmed

## Query development review checklist

For PRs that add new queries or modify existing queries, the following checklist should be completed by both the author and reviewer:

### Author

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### Reviewer 

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### CODE REVIEW

File: cpp/autosar/test/rules/A3-1-5/test.cpp [hunk 1]
```diff
@@ -23,7 +23,7 @@ class A {
```
Comment (knewbury01) [inline cpp/autosar/test/rules/A3-1-5/test.cpp:23]: no change note added yet to this PR

File: cpp/common/src/codingstandards/cpp/Class.qll [hunk 2]
```diff
@@ -192,7 +192,10 @@ class TrivialMemberFunction extends IntrospectedMemberFunction {
  * class.
  */
 class TemplateOrTemplateClassMemberFunction extends MemberFunction {
-  TemplateOrTemplateClassMemberFunction() { isFromUninstantiatedTemplate(_) }
+  TemplateOrTemplateClassMemberFunction() {
+    isFromUninstantiatedTemplate(_) or
+    isFromTemplateInstantiation(_)
```
Comment (knewbury01) [inline cpp/common/src/codingstandards/cpp/Class.qll:197]: issue #382 says we might want
```suggestion
(
      isFromUninstantiatedTemplate(_) or
      isFromTemplateInstantiation(_)
    ) and
    not this.isCompilerGenerated()
```

**Review by knewbury01 — CHANGES_REQUESTED** [review]
only small comments/changes, otherwise looks good!

File: cpp/autosar/test/rules/A3-1-5/test.cpp [hunk 3]
```diff
@@ -23,7 +23,7 @@ class A {
 
   int complexCalculation();
```
Comment (knewbury01) [inline cpp/autosar/test/rules/A3-1-5/test.cpp:25]: oh, also needs a test format

**Review by knewbury01 — CHANGES_REQUESTED** [review]
also, one more oddity, not sure how this file name was [passing our length check](https://github.com/github/codeql-coding-standards/actions/runs/9353792485/job/25745128639?pr=610#step:8:1708) before (ie it wasnt but also wasnt getting checked?)

not sure what to do about that part, either rename, or figure out while older query names are not/if they should be made to also comply with that

**Review by knewbury01 — COMMENTED** [review]
oh, @lcartey , one other thing.

perhaps consider fixing #611 at the same time? it is also for this query

possibly something like this? [here](https://github.com/github/codeql-coding-standards/blob/main/cpp/common/src/codingstandards/cpp/Class.qll#L151):
```
predicate hasTrivialLength() {
    this.getBlock().getLastStmt().getLocation().getStartLine() - this.getBlock().getStmt(0).getLocation().getStartLine() <= 10 and
    not exists(this.getBlock().getStmt(_).getChildStmt())
  }
```

File: cpp/autosar/test/rules/A3-1-5/test.cpp (hunk 1 above)
Comment (lcartey) [reply cpp/autosar/test/rules/A3-1-5/test.cpp:23]: Forgot to add, have included now.

File: cpp/common/src/codingstandards/cpp/Class.qll (hunk 2 above)
Comment (lcartey) [reply cpp/common/src/codingstandards/cpp/Class.qll:197]: For the current usage and query this isn't necessary - as we are using this class to _exclude_ cases, not include. As a belt-and-braces measure, we can add it in.

**Comment by lcartey** [comment]
@knewbury01 this is ready for your re-review

**Review by knewbury01 — APPROVED** [review]
nice work addressing previous comments, good looking/straightforward PR!
